spec/integration/indexer/indexing_spec.rb

```diff
@@ -180,9 +188,13 @@
         { "index" => { "error" => { "reason" => "[FORBIDDEN/metasearch/index read-only" } } }
       ] }
 
-      expect(stubbed_client).to receive(:bulk).and_return(locked_response)
-      expect(stubbed_client).to receive(:bulk).and_call_original
-      allow_any_instance_of(SearchIndices::Index).to receive(:build_client).and_return(stubbed_client)
+      index = SearchConfig.instance.search_server.index_group(index_name).current
+      index.lock
+
+      Thread.new do
+        sleep 0.1
```

Comment:
Why doesn't the original way of stubbing the response work anymore?

Reply:
`any_instance_of` works great when we expect only one instance of a class to receive the messages. However, when we want multiple classes to be called (all instances of this class should be called once, and in total twice), there's a gotcha.

From [rspec docs](https://relishapp.com/rspec/rspec-mocks/docs/working-with-legacy-code/any-instance):

> This feature is sometimes useful when working with legacy code, though in general we
discourage its use for a number of reasons:
> The rspec-mocks API is designed for individual object instances, but this feature operates on entire classes of objects. As a result there are some semantically confusing edge cases. For example, in expect_any_instance_of(Widget).to receive(:name).twice it isn't clear whether a specific instance is expected to receive name twice, or if two receives total are expected. (It's the former.)

This means that `allow_any_instance_of(SearchIndices::Index).to receive(:build_client)` wouldn't work with `.twice` (it'd expect each instance to get called twice).

As this is an integration test (similar to the worker specs), and `allow_any_instance_of` is discouraged, I rewrote the test to more closely mirror the index being locked.

Reply:
I'm not sure this is reliably testing the desired behaviour anymore, because there's no guarantee that the `post` will be rejected by the lock (and so be rescheduled) - the lock could expire before the `post` happens, in which case all we're testing is that we can insert a document by `post`ing.

Reply:
You're right, I could set the delay to way longer, but that still doesn't guarantee that the lock causes a retry. We only check that a lock happened and that the request was eventually successful.

One option would be to test it with only only active cluster?

```
allow(Clusters).to receive(:active).and_return([Clusters.default_cluster])
```

Then I could put back all the `allow_any_instance_of` stuff (:cry:). What do you think?

Reply:
I think that's probably the best solution

Reply:
Done